core/shared/src/main/scala/cats/parse/Accumulator.scala

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2020 Typelevel
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package cats.parse
+
+import cats.data.{NonEmptyList, NonEmptyVector}
+import scala.collection.mutable.Builder
+
+trait Appender[-A, +B] {
+  def append(item: A): this.type
+  def finish(): B
+}
+
+object Appender {
+  def charStringAppender(): Appender[Char, String] =
+    new Appender[Char, String] {
+      val bldr = new java.lang.StringBuilder()
+
+      def append(item: Char) = {
+        bldr.append(item)
+        this
+      }
+
+      def finish(): String = bldr.toString()
+    }
+
+  def stringAppender(): Appender[String, String] =
+    new Appender[String, String] {
+      val bldr = new java.lang.StringBuilder()
+
+      def append(item: String) = {
+        bldr.append(item)
+        this
+      }
+
+      def finish(): String = bldr.toString()
+    }
+
+  def fromBuilder[A, B](bldr: Builder[A, B]): Appender[A, B] =
+    new Appender[A, B] {
+      def append(item: A) = {
+        bldr += item
+        this
+      }
+
+      def finish(): B = bldr.result()
+    }
+}
+
+trait Accumulator[-A, +B] {
+  def newAppender(): Appender[A, B]
+}
+
+object Accumulator {
+  implicit val charStringAccumulator: Accumulator[Char, String] =
+    new Accumulator[Char, String] {
+      def newAppender() = Appender.charStringAppender()
+    }
+
+  implicit val stringAccumulator: Accumulator[String, String] =
+    new Accumulator[String, String] {
+      def newAppender() = Appender.stringAppender()
+    }
+
+  implicit def listAccumulator[A]: Accumulator[A, List[A]] =
```

Comment:
Should we have something like:

```scala
def fromBuilder[A, B](makeBuilder: () => Builder[A, B]): Accumulator[A, B] = ???
```

You could use that for both `listAccumulator` and `vectorAccumulator` and anyone else who wants to use `Appender.fromBuilder` will also want something like this.

Reply:
Also what about:

```scala
implicit def monoidAccumulator[A](implicit ev: Monoid[A]): Accumulator[A, A] = ???
```

(And the same for `Semigroup` with `Accumulator1`.)

Reply:
Lastly, I can imagine arguments for and against including `Set`. The argument for is that often users may want to end up with a set, and if their language ignores duplicates the behavior will be correct. The argument against is that information would be lost, and maybe the user should have to do that themselves. I could imagine building a `MultiSet[A]` (e.g. `Map[A, Int]`) although that seems kinds of loony from a documentation standpoint.

Reply:
These are good questions, but I left them undone for now:

1. I think most use cases will be the ones I made, and I didn't want to add an indirection through one more lambda.
2. As we talked offline, I can't see a good use case and plenty of weird cases for Monoid or Semigroup. For now, I'd rather punt till we have clearer examples. I don't know if it will be much more efficient than mapping and using combine by hand, which also has the benefit of reading more clearly